python/private/platform_runtime_repo.bzl

```diff
@@ -0,0 +1,196 @@
+# Copyright 2024 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Create a repository to hold a local Python toolchain."""
+
+load("//python/private:toolchains_repo.bzl", "get_host_os_arch")
+load(":repo_utils.bzl", "REPO_DEBUG_ENV_VAR", "repo_utils")
+
+_TOOLCHAIN_IMPL_TEMPLATE = """\
+# Generated by python/private/platform_runtime_repo.bzl
+
+load("@rules_python//python:py_runtime.bzl", "py_runtime")
+load("@rules_python//python:py_runtime_pair.bzl", "py_runtime_pair")
+load("@rules_python//python/cc:py_cc_toolchain.bzl", "py_cc_toolchain")
+load("@rules_python//python/private:py_exec_tools_toolchain.bzl", "py_exec_tools_toolchain")
+
+cc_library(
+    name = "python_headers",
+    srcs = glob(["include/**/*.h"]),
+    includes = ["include"],
+)
+
+cc_library(
+    name = "libpython",
+    # Don't use a recursive glob because the lib/ directory usually contains
+    # a subdirectory of the stdlib -- lots of unrelated files
+    srcs = glob([
+        "lib/*{lib_ext}",  # Match libpython*.so
+        "lib/*{lib_ext}*",  # Also match libpython*.so.1.0
+    ]),
+    hdrs = [":python_headers"],
+)
+
+py_runtime(
+    name = "py3_runtime",
+    interpreter_path = "{interpreter_path}",
+    python_version = "PY3",
+    interpreter_version_info = {{
+        "major": "{major}",
+        "minor": "{minor}",
+        "micro": "{micro}",
+    }},
+    implementation_name = "{implementation_name}",
+)
+
+py_runtime_pair(
+    name = "python_runtimes",
+    py2_runtime = None,
+    py3_runtime = ":py3_runtime",
+)
+
+py_exec_tools_toolchain(
+    name = "py_exec_tools_toolchain",
+)
+
+py_cc_toolchain(
+    name = "py_cc_toolchain",
+    headers = ":python_headers",
+    libs = ":libpython",
+    python_version = "{major}.{minor}.{micro}",
+)
+"""
+
+_GET_INFO_PROGRAM = """
+import sys
+import sysconfig
+import json
+
+data = {
+  "major": sys.version_info.major,
+  "minor": sys.version_info.minor,
+  "micro": sys.version_info.micro,
+  "include": sysconfig.get_path("include"),
+  "implementation_name": sys.implementation.name,
+}
+
+config_vars = [
+    "LDLIBRARY", "LIBDIR", "INSTSONAME", "LIBDEST", "PY3LIBRARY",
+    "SHLIB_SUFFIX",
+]
+data.update(zip(config_vars, sysconfig.get_config_vars(*config_vars)))
+print(json.dumps(data))
+"""
+
+def _platform_runtime_repo_impl(rctx):
```

Comment:
nit, it would be great to gate this to only work on specific OSes, e.g. check at the top if we are running on Windows and bail if not.

Reply:
Aside from the symlink on the include dir, is there something else in here that won't work on windows?

Honestly, I have no idea how correct/accurate/portable the code for find the include dir etc is. I couldn't find much docs about this, so reverse engineers what I could find on SO and what I saw on my linux laptop and desktop. No idea how things look on a mac or windows.

Reply:
We could probably use the CI what breaks. My thinking was to be on the safe side and enable only the platforms that are known to work. That way we can segment the problem into smaller chunks.